### PR TITLE
DM-34874: Refactor parquet formatter to use Arrow Tables natively, and add converters.

### PR DESCRIPTION
**Pull request by erykoff**
## Checklist

- [x] ran Jenkins
- [x] added a release note for user-visible changes to `doc/changes`

### CODE REVIEW

File: python/lsst/daf/butler/configs/datastores/formatters.yaml [hunk 1]
```diff
@@ -71,7 +71,9 @@ StampsBase: lsst.obs.base.formatters.fitsGeneric.FitsGenericFormatter
 Stamps: lsst.obs.base.formatters.fitsGeneric.FitsGenericFormatter
 BrightStarStamps: lsst.obs.base.formatters.fitsGeneric.FitsGenericFormatter
 AstropyTable: lsst.daf.butler.formatters.astropyTable.AstropyTableFormatter
-AstropyQTable: lsst.daf.butler.formatters.astropyTable.AstropyTableFormatter
```
Comment (timj) [inline python/lsst/daf/butler/configs/datastores/formatters.yaml:74]: Why is this removed?

File: python/lsst/daf/butler/delegates/arrowastropy.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+        AttributeError
+            The component can not be found.
+        """
+        if componentName == "columns":
+            return list(composite.columns.keys())
+        elif componentName == "schema":
+            return ArrowAstropySchema(composite)
+        elif componentName == "rowcount":
+            return len(composite)
+        else:
```
Comment (timj) [inline python/lsst/daf/butler/delegates/arrowastropy.py]: Doesn't need the else here since you want to always raise if you get to the end of the method.

File: python/lsst/daf/butler/delegates/arrowastropy.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+        -------
+        component : `object`
+            The component.
+
+        Raises
+        ------
+        AttributeError
+            The component can not be found.
+        """
+        if componentName == "columns":
```
Comment (timj) [inline python/lsst/daf/butler/delegates/arrowastropy.py]: Can we use the new match/case here?

File: python/lsst/daf/butler/delegates/arrownumpy.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+        elif componentName == "schema":
+            return ArrowNumpySchema(composite.dtype)
+        elif componentName == "rowcount":
+            return len(composite)
+        else:
+            raise AttributeError(
+                f"Do not know how to retrieve component {componentName} from {get_full_type_name(composite)}"
+            )
+
+    def handleParameters(
```
Comment (timj) [inline python/lsst/daf/butler/delegates/arrownumpy.py]: There's an uncomfortable amount of code duplication between this method and the ones in the other delegates. Seems like only one line differs (the "if column not in") and the instance check (the python type for that could be a class attribute). I don't like that much repeated logic across multiple files.

File: python/lsst/daf/butler/formatters/parquet.py [hunk 5]
```diff
@@ -21,177 +21,842 @@
...
+            use_threads=False,
+            use_pandas_metadata=(b"pandas" in metadata),
+            filters=filters,
+        )
+
+        return arrow_table
+
+    def write(self, inMemoryDataset: Any) -> None:
+        import numpy as np
+        import pandas as pd
```
Comment (timj) [inline python/lsst/daf/butler/formatters/parquet.py]: Importing pandas just to check if we have been given a pandas dataframe is a tad annoying.

File: python/lsst/daf/butler/formatters/parquet.py [hunk 6]
```diff
@@ -21,177 +21,842 @@
...
+        from astropy.table import Table as astropyTable
+
+        location = self.makeUpdatedLocation(self.fileDescriptor.location)
```
Comment (timj) [inline python/lsst/daf/butler/formatters/parquet.py]: move this line down to where it's needed (just before the `write_table`)

File: python/lsst/daf/butler/formatters/parquet.py [hunk 7]
```diff
@@ -21,177 +21,842 @@
...
+            arrow_table = astropy_to_arrow(inMemoryDataset)
+        elif isinstance(inMemoryDataset, np.ndarray):
+            arrow_table = numpy_to_arrow(inMemoryDataset)
+        elif isinstance(inMemoryDataset, pa.Table):
+            arrow_table = inMemoryDataset
         else:
-            raw_columns = list(self._splitColumnnNames(len(indexes), self.file.metadata.schema.names))
-            self.columns = pd.MultiIndex.from_tuples(raw_columns, names=[f["name"] for f in indexes])
-        self.indexLevelNames = tuple(self.columns.names)
+            raise ValueError("Unsupported type of inMemoryDataset for ParquetFormatter.")
```
Comment (timj) [inline python/lsst/daf/butler/formatters/parquet.py]: Maybe report the get_full_type_name in this message?

File: python/lsst/daf/butler/formatters/parquet.py [hunk 8]
```diff
@@ -21,177 +21,842 @@
...
+
+        pq.write_table(arrow_table, location.path)
+
+
+def arrow_to_pandas(arrow_table: pa.Table) -> pd.DataFrame:
+    """Convert a pyarrow table to a pandas DataFrame.
+    If the input table has ``pandas`` metadata in the schema it will be
```
Comment (timj) [inline python/lsst/daf/butler/formatters/parquet.py]: Blank line before this one since this is a note and not the short summary. It should probably be attached to the parameter description below.

File: python/lsst/daf/butler/formatters/parquet.py [hunk 9]
```diff
@@ -21,177 +21,842 @@
...
+def arrow_to_numpy(arrow_table: pa.Table) -> np.ndarray:
+    """Convert a pyarrow table to a structured numpy array.
+
+    Parameters
+    ----------
+    arrow_table : `pyarrow.Table`
+
+    Returns
+    -------
+    array : `numpy.ndarray` (N,)
```
Comment (timj) [inline python/lsst/daf/butler/formatters/parquet.py:212]: Maybe add a short description of the shape of the numpy array here rather than using `N` where `N` has no context.

File: python/lsst/daf/butler/formatters/parquet.py [hunk 10]
```diff
@@ -21,177 +21,842 @@
...
+
+
+def arrow_to_pandas(arrow_table: pa.Table) -> pd.DataFrame:
+    """Convert a pyarrow table to a pandas DataFrame.
+    If the input table has ``pandas`` metadata in the schema it will be
+    used in the construction of the DataFrame.
+
+    Parameters
+    ----------
+    arrow_table : `pyarrow.Table`
```
Comment (timj) [inline python/lsst/daf/butler/formatters/parquet.py:166]: Normally we'd have a short text of something like "The table to be converted." rather than no description at all (applies to all these parameters).

File: python/lsst/daf/butler/formatters/parquet.py [hunk 11]
```diff
@@ -21,177 +21,842 @@
...
+    dataframe : `pandas.DataFrame`
+        Dataframe to turn into a schema.
+    """
+
+    def __init__(self, dataframe: pd.DataFrame) -> None:
+        self._schema = dataframe.loc[[False] * len(dataframe)]
+
+    @classmethod
+    def from_arrow(cls, schema: pa.Schema) -> DataFrameSchema:
+        """Convert an arrow schema into a DataFrameSchema.
```
Comment (timj) [inline python/lsst/daf/butler/formatters/parquet.py]: backticks around `DataFrameSchema`?

File: python/lsst/daf/butler/formatters/parquet.py [hunk 12]
```diff
@@ -21,177 +21,842 @@
...
-            Number of levels in the `pd.MultiIndex` that is being
-            reconstructed.
-        names : `~collections.abc.Iterable` of `str`
-            Strings to be split.
-
-        Yields
-        ------
-        tuple : `tuple` of `str`
-            A multi-index column name tuple.
+        schema : `pyarrow.Schema`
```
Comment (timj) [inline python/lsst/daf/butler/formatters/parquet.py:479]: Short description?

File: python/lsst/daf/butler/formatters/parquet.py [hunk 13]
```diff
@@ -21,177 +21,842 @@
...
+        Returns
+        -------
+        arrow_schema : `pyarrow.Schema`
+        """
+        arrow_table = pa.Table.from_pandas(self._schema)
+
+        return arrow_table.schema
+
+    def to_arrow_numpy_schema(self) -> ArrowNumpySchema:
+        """Convert to an ArrowNumpySchema.
```
Comment (timj) [inline python/lsst/daf/butler/formatters/parquet.py]: backticks?

File: python/lsst/daf/butler/formatters/parquet.py [hunk 14]
```diff
@@ -21,177 +21,842 @@
...
+        if not isinstance(other, ArrowAstropySchema):
+            return False
+
+        # If this comparison passes then the two tables have the
+        # same column names.
+        if not self._schema.dtype == other._schema.dtype:
+            return False
+
+        for name in self._schema.columns:
+            if not self._schema[name].unit == other._schema[name].unit:
```
Comment (timj) [inline python/lsst/daf/butler/formatters/parquet.py:621]: Am I missing something subtle as to why this isn't using `!=`?

File: python/lsst/daf/butler/formatters/parquet.py [hunk 15]
```diff
@@ -21,177 +21,842 @@
...
+    @property
+    def schema(self) -> atable.Table:
+        return self._schema
+
+    def __repr__(self) -> str:
+        return repr(self._schema)
+
+    def __eq__(self, other: object) -> bool:
+        if not isinstance(other, ArrowAstropySchema):
+            return False
```
Comment (timj) [inline python/lsst/daf/butler/formatters/parquet.py]: `return NotImplemented` ?

File: python/lsst/daf/butler/formatters/parquet.py [hunk 16]
```diff
@@ -21,177 +21,842 @@
...
+    ----------
+    numpy_dtype : `numpy.dtype` or `list` [`str`]
+    """
+
+    def __init__(self, numpy_dtype: np.dtype) -> None:
+        self._dtype = numpy_dtype
+
+    @classmethod
+    def from_arrow(cls, schema: pa.Schema) -> ArrowNumpySchema:
+        """Convert an arrow schema into an ArrowNumpySchema.
```
Comment (timj) [inline python/lsst/daf/butler/formatters/parquet.py]: back ticks.

File: python/lsst/daf/butler/formatters/parquet.py [hunk 17]
```diff
@@ -21,177 +21,842 @@
...
+        return True
+
+
+class ArrowNumpySchema:
+    """Wrapper class for a schema for a numpy ndarray.
+
+    Parameters
+    ----------
+    numpy_dtype : `numpy.dtype` or `list` [`str`]
+    """
```
Comment (timj) [inline python/lsst/daf/butler/formatters/parquet.py:638]: parameter description text?

File: python/lsst/daf/butler/formatters/parquet.py [hunk 18]
```diff
@@ -21,177 +21,842 @@
...
-    pq.write_table(table, path)
+            dtype.append((name, _arrow_string_to_numpy_dtype(schema, name)))
 
+        return cls(np.dtype(dtype))
 
-class ParquetFormatter(Formatter):
-    """Interface for reading and writing Pandas DataFrames to and from Parquet
-    files.
+    def to_arrow_astropy_schema(self) -> ArrowAstropySchema:
+        """Convert to an ArrowAstropySchema.
```
Comment (timj) [inline python/lsst/daf/butler/formatters/parquet.py]: Backticks? (I'll stop now since there are more examples)

File: python/lsst/daf/butler/formatters/parquet.py [hunk 19]
```diff
@@ -21,177 +21,842 @@
...
+        for requested in columns:
+            if not isinstance(requested, tuple):
+                raise ValueError(
+                    "Columns parameter for multi-index data frame must be a " "dictionary or list of tuples."
+                )
+            names.append(str(requested))
+    else:
+        if not isinstance(columns, collections.abc.Mapping):
+            raise ValueError(
+                "Columns parameter for multi-index data frame must be a " "dictionary or list of tuples."
```
Comment (timj) [inline python/lsst/daf/butler/formatters/parquet.py]: Say what type you got?

File: python/lsst/daf/butler/formatters/parquet.py [hunk 20]
```diff
@@ -21,177 +21,842 @@
...
+    return names
+
+
+def _apply_astropy_metadata(astropy_table: atable.Table, metadata: Dict) -> None:
+    """Apply any astropy metadata from the schema metadata.
+
+    Parameters
+    ----------
+    astropy_table : `astropy.table.Table`
+        Table to apply metdata
```
Comment (timj) [inline python/lsst/daf/butler/formatters/parquet.py]: typo: metdata

File: python/lsst/daf/butler/formatters/parquet.py [hunk 21]
```diff
@@ -21,177 +21,842 @@
...
+    Returns
+    -------
+    dtype_str : `str`
+        Numpy dtype string.
+    """
+    # Special-case for string and binary columns
+    md_name = f"lsst::arrow::len::{name}"
+    strlen = default_length
+    metadata = schema.metadata if schema.metadata is not None else {}
+    if md_name.encode("UTF-8") in metadata:
```
Comment (timj) [inline python/lsst/daf/butler/formatters/parquet.py]: can you say:

```python
if (encoded := md_name.encode("UTF-8")) in metadata:
    strlen = int(schema.metadata[encoded])
```
so you don't need to encode twice?

File: python/lsst/daf/butler/formatters/parquet.py [hunk 22]
```diff
@@ -21,177 +21,842 @@
...
+        Column name.
+    dtype : `np.dtype`
+        Numpy dtype.
+    """
+    import numpy as np
+
+    if dtype.type is np.str_:
+        metadata[f"lsst::arrow::len::{name}".encode("UTF-8")] = str(dtype.itemsize // 4)
+    elif dtype.type is np.bytes_:
+        metadata[f"lsst::arrow::len::{name}".encode("UTF-8")] = str(dtype.itemsize)
```
Comment (timj) [inline python/lsst/daf/butler/formatters/parquet.py:886]: Is this called with a non-string dtype ever?

**Review by timj — APPROVED** [review]
Some minor cleanups noted and I worry about code duplication.

Thanks for the comprehensive tests with excellent code coverage.

File: python/lsst/daf/butler/configs/datastores/formatters.yaml (hunk 1 above)
Comment (erykoff) [reply python/lsst/daf/butler/configs/datastores/formatters.yaml:74]: I don't know.  Restoring it.

File: python/lsst/daf/butler/delegates/arrowastropy.py (hunk 3 above)
Comment (erykoff) [reply python/lsst/daf/butler/delegates/arrowastropy.py]: Sure, why not?

File: python/lsst/daf/butler/formatters/parquet.py (hunk 22 above)
Comment (erykoff) [reply python/lsst/daf/butler/formatters/parquet.py:886]: Yes, and then this is a no-op.  Is that a problem?

Comment (timj) [reply python/lsst/daf/butler/formatters/parquet.py:886]: I wanted to know if the dangling else was okay. Maybe add a note to the docstring saying that all column types are handled but only string types are modified?
